core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala

```diff
@@ -269,4 +302,22 @@ object Scheduler {
       }
     }
   }
+
+  val PipeliningLimit = 128
+
+  private[core] def persistenceFlow(
+      podRecordRepository: PodRecordRepository
+  )(implicit materializer: Materializer): Flow[SchedulerEvents, SchedulerEvents, NotUsed] = {
+    Flow[SchedulerEvents].mapAsync(1) { events =>
+      Source(events.stateEvents).collect {
+        case PodRecordUpdated(_, Some(podRecord)) => podRecordRepository.store(podRecord)
+        case PodRecordUpdated(podId, None) => podRecordRepository.delete(podId)
+      }.mapAsync(PipeliningLimit)(identity)
```

Comment:
Please excuse my ignorance I probably miss some context.

First off, is `Source(...).collect...mapAsync.runWith` returning a future? I don't get how `Flow[SchedulerEvents].mapAsync(1)` would work otherwise. Could we create a simple helper function:

```scala
def presistChanges(SchedulerEvents): Future[???] = {
  Source(events.stateEvents).collect {
    case PodRecordUpdated(_, Some(podRecord)) => podRecordRepository.store(podRecord)
    case PodRecordUpdated(podId, None) => podRecordRepository.delete(podId)
  }
  .mapAsync(PipeliningLimit)(identity)
  .runWith(Sink.ignore)
  .map(_ => events)(CallerThreadExecutionContext.context)
}

Flow[SchedulerEvents].mapAsync(1)(persistChanges)
```

Second, on first sight it seems we are calling `store` and `delete` for all `stateEvents` and then limit them. I expect Akka to only make the call when `mapAsync` pulls the next items. Is that correct @timcharper? What do you think about making this more explicit by providing the call as anonymous functions?

```scala
def presistChanges(SchedulerEvents): Future[???] = {
  Source(events.stateEvents).collect {
    case PodRecordUpdated(_, Some(podRecord)) => () => podRecordRepository.store(podRecord)
    case PodRecordUpdated(podId, None) => () => podRecordRepository.delete(podId)
  }
  .mapAsync(PipeliningLimit) { call => call() }
  .runWith(Sink.ignore)
  .map(_ => events)(CallerThreadExecutionContext.context)
}
```

Reply:
After discussing this with @timcharper , we refactored the code to apply Pipelining limit across multiple `SchedulerEvents` objects, i have also applied @jeschkies suggestion on using `() => Future[..]` instead of `Future[..]`. The new approach uses Future's but restricts them with an upper limit. Added a test case to verify the same.